proposals/kapp/001-crd-upgrade-safety/README.md

```diff
@@ -0,0 +1,95 @@
+
+As an example, configuring a preflight check could look something like:
+```sh
+kapp deploy ... --preflight=Check=true:{"mode":"strict"}
+```
+
+To provide a high quality user experience, each configurable preflight check should have a sane default and not require a user to provide any configuration information.
+
+#### `CRDUpgradeSafety` Logic
+Updates to a CRD can be deemed unsafe if any of the following occur:
+- An existing version that is a stored version in the object store is removed
+- Scope of a CRD is changed from Namespace to Cluster (or vice versa)
+- Removing a field in an existing version
+- Adding a new required field
+- Updates to an existing field that makes validation more restrictive or would break existing versions/clients. Some brief examples: changing a field's type from string to int, increasing minimum value, removing an enum value.
+
+The most straightforward changes to check for are the first four and will have explicit checks in the implementation.
+
+Checking for breaking changes to an existing field is much more nuanced and is more difficult to catch all valid cases.
+It is proposed that an iterative approach is taken where the base implementation does not contain any specific checks against changed fields, but is written in a way that it is easy to add checks in the future.
+This does not mean that we shouldn't add any specific checks against changed fields before a release of `kapp` containing this preflight check, but rather focus on laying a foundation in which we can iteratively build on.
+Some potential field checks we could add after the base implementation is in place are:
+- Type is not changed
+- Enum value is not removed
+- Minimum value is not increased
+- Maximum value is not decreased
+
+To start with, there will be a couple configuration options for this preflight check:
+- **mode** - allowed values: ["error", "warn"] - This is the mode that should be used when running the preflight check. When run in the `error` mode, this check will fail if any problems are found with upgrading a CRD and prevent the changes from being applied to the cluster. When run in `warn` mode, this check will always succeed, but log warnings for any problems found. The `warn` mode allows users to be aware of potential problems but proceed with applying the changes to the cluster if all other enabled preflight checks pass. `error` is the default.
+- **failMode** - allowed values: ["open", "closed"] - This is the failure mode that should be used when running the preflight check. Only used when `mode=error`. When running in the `open` mode, if a change to an existing field is not validated by an existing check the preflight check will pass. When running in the `closed` mode, all changes to an existing field *must* be validated by an existing check. If it has not, the preflight check will fail with a message along the lines of "unable to determine if change to existing field is safe ...". `closed` is the default.
```

Comment:
I might suggest that `open` be the default, as it maintains existing behavior. `closed` is fairly strict, and might lead to many more failures than expected.

Reply:
I think it's okay for `closed` to be the default here as this is an opt-in preflight check. The default behavior should be the same as the preflight check is disabled by default.